llvm/lib/Target/Xtensa/XtensaISelLowering.cpp

```diff
@@ -827,47 +827,8 @@ SDValue XtensaTargetLowering::LowerShiftRightParts(SDValue Op,
 }
 
 SDValue XtensaTargetLowering::LowerCTPOP(SDValue Op, SelectionDAG &DAG) const {
-  EVT VT = Op->getValueType(0);
-  SDValue Val = Op.getOperand(0);
-  SDLoc DL(Op);
-
-  if (VT != MVT::i32)
-    return SDValue();
-
-  // CTPOP expansion:
-  // Val = (Val - (Val >> 1)) & 0x55555555
-  // Val = ((Val >> 2) & 0x33333333) + (Val & 0x33333333)
-  // Val = ((Val >> 4) + Val) & 0x0f0f0f0f
-  // Val = (Val >> 8) + Val
-  // Val = (extract bits [16, 20] from Val) + Val
-  // Val = extract bits [0, 5] from Val
-
-  SDValue Mask = DAG.getConstant(0x55555555, DL, VT);
-  SDValue Shift =
-      DAG.getNode(ISD::SRL, DL, VT, Val, DAG.getConstant(1, DL, VT));
-  SDValue ShiftAndMask = DAG.getNode(ISD::AND, DL, VT, Shift, Mask);
-  Val = DAG.getNode(ISD::SUB, DL, VT, Val, ShiftAndMask);
-
-  Mask = DAG.getConstant(0x33333333, DL, VT);
-  Shift = DAG.getNode(ISD::SRL, DL, VT, Val, DAG.getConstant(2, DL, VT));
-  SDValue ValAndMask = DAG.getNode(ISD::AND, DL, VT, Val, Mask);
-  ShiftAndMask = DAG.getNode(ISD::AND, DL, VT, Shift, Mask);
-  Val = DAG.getNode(ISD::ADD, DL, VT, ValAndMask, ShiftAndMask);
-
-  Mask = DAG.getConstant(0x0f0f0f0f, DL, VT);
-  Shift = DAG.getNode(ISD::SRL, DL, VT, Val, DAG.getConstant(4, DL, VT));
-  Val = DAG.getNode(ISD::ADD, DL, VT, Val, Shift);
-  Val = DAG.getNode(ISD::AND, DL, VT, Val, Mask);
-
-  Shift = DAG.getNode(ISD::SRL, DL, VT, Val, DAG.getConstant(8, DL, VT));
-  Val = DAG.getNode(ISD::ADD, DL, VT, Val, Shift);
-
-  Shift = DAG.getNode(XtensaISD::EXTUI, DL, VT, Val,
-                      DAG.getConstant(16, DL, VT), DAG.getConstant(5, DL, VT));
-  Val = DAG.getNode(ISD::ADD, DL, VT, Val, Shift);
-
-  return DAG.getNode(XtensaISD::EXTUI, DL, VT, Val, DAG.getConstant(0, DL, VT),
-                     DAG.getConstant(6, DL, VT));
+  auto &TLI = DAG.getTargetLoweringInfo();
+  return TLI.expandCTPOP(Op.getNode(), DAG);
```

Comment:
TLI is this. Also I would expect this custom lowering to just be identical to what you get for expand?

Reply:
> TLI is this. Also I would expect this custom lowering to just be identical to what you get for expand?

The instruction sequence which I proposed before is a bit different from sequence generated by  TLI.expandCTPOP(). It is because I used "right shift and mask" at the last stage and thus need to use 2 EXTUI operations in my implementation. The TLI.expandCTPOP() uses "left shift and mask" at the last stage, as I understand, but instruction amount in both implementations is the same.

Reply:
Also CTPOP isn't mentioned in the patch description. Should this be done separately?

Reply:
You mean to change first patch description from "[Xtensa] Implement lowering Mul/Div/Shift operations." to something like "[Xtensa] Implement lowering Mul/Div/Shift/CTTZ/CTLZ/CTPOP operations."? I can force update patches

Reply:
Either that, or split the ctpop into a separate pR